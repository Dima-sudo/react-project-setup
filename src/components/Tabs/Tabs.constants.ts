/* eslint-disable no-unused-vars */
export enum DataTestIds {
    COMPONENT_CONTAINER = 'tabs-component__container',
}
