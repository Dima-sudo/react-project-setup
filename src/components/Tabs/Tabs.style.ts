import { styled } from '@mui/material/styles';

import TabsComponent from '@mui/material/Tabs';
import Box from '@mui/material/Box';

export const Tabs = styled(TabsComponent)(({ color }) => ({
    '& button.Mui-selected': {
        color: color || 'unset',
    },
}));

export const ComponentContainer = styled(Box, {
    shouldForwardProp: (prop) => prop !== 'tabsBackgroundColor',
})<{ tabsBackgroundColor?: string; variant?: string }>(
    ({
        theme: {
            custom: { colors, params, shadows },
        },
        tabsBackgroundColor,
        variant,
    }) => ({
        backgroundColor:
            variant !== 'transparent'
                ? tabsBackgroundColor || colors.white
                : 'transparent',
        borderRadius: params.borderRadius,
        boxShadow:
            variant !== 'transparent' ? shadows.boxShadows.boxShadow1 : 'unset',
        transition: params.transitionDefault,

        '&:hover': {
            boxShadow:
                variant !== 'transparent'
                    ? shadows.boxShadows.boxShadow2
                    : 'unset',
            transition: params.transitionDefault,
        },
    })
);
