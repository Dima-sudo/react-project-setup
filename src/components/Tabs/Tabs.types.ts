import React from 'react';

export interface TabPanelProps {
    children?: React.ReactNode | React.ReactElement;
    dir?: string;
    index: number;
    value: number;
}

export interface ITab {
    label: string;
    content: React.ReactNode | React.ReactElement;
    isDisabled?: boolean;
    dataTestId?: string;
}
