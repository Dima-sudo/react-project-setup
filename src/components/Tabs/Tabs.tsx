import React, { useState, useCallback, SyntheticEvent } from 'react';

import { useTheme } from '@mui/material/styles';
import Tab from '@mui/material/Tab';
import SwipeableViews from 'react-swipeable-views';

import TabPanel from './components/TabPanel/TabPanel';

import { HexColor } from '../../types/common.types';
import { ITab } from './Tabs.types';
import { DataTestIds } from './Tabs.constants';
import * as S from './Tabs.style';

interface IProps {
    tabs: ITab[];
    color?: HexColor;
    tabsBackgroundColor?: HexColor;
    variant?: 'transparent' | 'primary';
    layout?: 'standard' | 'fullWidth' | 'scrollable';
    className?: string;
    ariaLabel?: string;
}

const Tabs: React.FC<IProps> = ({
    tabs,
    color,
    tabsBackgroundColor,
    variant = 'primary',
    layout = 'standard',
    className,
    ariaLabel = 'aria-tabs-component',
    ...rest
}) => {
    const theme = useTheme();
    const [selectedIndex, setSelectedIndex] = useState(0);

    const onChange = (event: SyntheticEvent, index: number) => {
        setSelectedIndex(index);
    };

    const onSwipeableChange = (index: number) => {
        setSelectedIndex(index);
    };

    const a11yProps = useCallback(
        (index: number) => ({
            id: `tab-${index}`,
            'aria-controls': `tabpanel-${index}`,
        }),
        []
    );

    return (
        <S.ComponentContainer
            {...{ tabsBackgroundColor, variant, className }}
            data-testid={DataTestIds.COMPONENT_CONTAINER}
        >
            <S.Tabs
                value={selectedIndex}
                onChange={onChange}
                variant={layout}
                aria-label={ariaLabel}
                TabIndicatorProps={{
                    sx: { backgroundColor: color || 'primary' },
                }}
                color={color}
                {...rest}
            >
                {tabs.map(({ label, isDisabled, dataTestId }, index) => (
                    <Tab
                        label={label}
                        key={label}
                        disabled={isDisabled}
                        data-testid={dataTestId || `tabs-tab__${index}`}
                        {...a11yProps(index)}
                    />
                ))}
            </S.Tabs>

            <SwipeableViews
                axis={theme.direction === 'rtl' ? 'x-reverse' : 'x'}
                index={selectedIndex}
                onChangeIndex={onSwipeableChange}
            >
                {tabs.map(({ content }, index) => (
                    <TabPanel
                        value={selectedIndex}
                        index={index}
                        dir={theme.direction}
                        key={`tab-content__${content?.toString().slice(0, 10)}`}
                    >
                        {content}
                    </TabPanel>
                ))}
            </SwipeableViews>
        </S.ComponentContainer>
    );
};

export default Tabs;
