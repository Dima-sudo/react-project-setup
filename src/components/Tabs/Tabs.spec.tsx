import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';

import Setup from '../../tests/setup/setup';
import { DataTestIds } from './Tabs.constants';
import Tabs from './Tabs';

describe('Tabs test suite', () => {
    it('Renders Tabs component correctly', () => {
        const tabs = [
            { label: 'tab1', content: <div>content1</div> },
            { label: 'tab2', content: <div>content2</div> },
        ];

        render(
            <Setup>
                <Tabs tabs={tabs} />
            </Setup>
        );

        const tabsComponent = screen.getByTestId(
            DataTestIds.COMPONENT_CONTAINER
        );

        expect(tabsComponent).toBeInTheDocument();
    });

    it('Snapshots component', () => {
        const tabs = [
            { label: 'tab1', content: <div>content1</div> },
            { label: 'tab2', content: <div>content2</div> },
        ];

        render(
            <Setup>
                <Tabs tabs={tabs} />
            </Setup>
        );

        const tabsComponent = screen.getByTestId(
            DataTestIds.COMPONENT_CONTAINER
        );

        expect(tabsComponent).toMatchSnapshot();
    });

    it('Recognizes disabled tabs', () => {
        const tabs = [
            { label: 'tab0', content: <div>content0</div> },
            { label: 'tab1', content: <div>content1</div>, isDisabled: true },
        ];

        render(
            <Setup>
                <Tabs tabs={tabs} />
            </Setup>
        );

        const disabledTab = screen.getByTestId('tabs-tab__1');

        expect(disabledTab).toBeDisabled();
    });
});
