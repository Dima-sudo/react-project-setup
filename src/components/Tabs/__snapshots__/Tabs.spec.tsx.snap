// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs test suite Snapshots component 1`] = `
<div
  class="MuiBox-root css-g43f0e"
  data-testid="tabs-component__container"
  variant="primary"
>
  <div
    class="MuiTabs-root css-1peytwk-MuiTabs-root"
  >
    <div
      class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
      style="overflow: hidden; margin-bottom: 0px;"
    >
      <div
        aria-label="aria-tabs-component"
        class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
        role="tablist"
      >
        <button
          aria-controls="tabpanel-0"
          aria-selected="true"
          class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
          data-testid="tabs-tab__0"
          id="tab-0"
          role="tab"
          tabindex="0"
          type="button"
        >
          tab1
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-controls="tabpanel-1"
          aria-selected="false"
          class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
          data-testid="tabs-tab__1"
          id="tab-1"
          role="tab"
          tabindex="-1"
          type="button"
        >
          tab2
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <span
        class="MuiTabs-indicator css-19zayo3-MuiTabs-indicator"
        style="left: 0px; width: 0px;"
      />
    </div>
  </div>
  <div
    style="overflow-x: hidden;"
  >
    <div
      class="react-swipeable-view-container"
      style="flex-direction: row; transition: all 0s ease 0s; direction: ltr; display: flex; will-change: transform;"
    >
      <div
        aria-hidden="false"
        data-swipeable="true"
        style="width: 100%; flex-shrink: 0; overflow: auto;"
      >
        <div
          aria-labelledby="tab-0"
          data-testid="tabs-tabpanel__0"
          dir="ltr"
          id="tabpanel-0"
          role="tabpanel"
        >
          <div
            class="MuiBox-root css-19kzrtu"
          >
            <div>
              content1
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
